BJS2-65053

gwalka opened the pull request:
решение задачи

Commit:
BJS2-65053
решение задачи

yom-yoom commented on src/main/java/school/faang/user_service/filter/user/UserFilter.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.stream.Stream;
+
+public interface UserFilter {
+    public boolean isApplicable(UserFilterDto filter);
```
Тут модификатор public необязателен

yom-yoom commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,78 @@
+            throw new DataValidationException("%d уже подписан на %d.", followerId, followeeId);
+        } else {
+            repository.followUser(followerId, followeeId);
+            log.info(String.format("Пользователь: %d подписался на пользователя: %d.", followerId, followeeId));
```
Использование String.format в логах не является хорошей практикой из за потери некоторой производительности, вместо этого можно передать аргументы через {} внутри кавычек, например: log.info("parameter: {}", followerId);

yom-yoom commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,78 @@
+
+    public void unfollowUser(long followerId, long followeeId) {
+        repository.unfollowUser(followerId, followeeId);
+        log.info(String.format("Пользователь: %d отписался от пользователя: %d.", followerId, followeeId));
```
Тут тоже

yom-yoom commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,78 @@
+        log.info(String.format("Пользователь: %d отписался от пользователя: %d.", followerId, followeeId));
+    }
+
+    public List<UserDto> getFollowers(long followeedId, UserFilterDto filter) {
```
followeed опечатка)

yom-yoom commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,4 @@
+package school.faang.user_service.dto;
+
+public record UserDto(long id, String username, String email) {
```
Круто, что используешь record)

zaqqaz226 replied:
Присоединяюсь!)

Commit:
String.format убран
опечатка follooweed исправлена

Commit:
public booelan == boolean

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,52 @@
+    private final SubscriptionService subscriptionService;
+
+    public void followUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
```
С точки зрения правильного дизайна приложения, контроллер, это чать, которая принимает запросы и отдает ответы и все. Те здесь не должно быть какой-то бизнес логики. В данном случае, проверка больше относится именно к бизнес логике. Так что лучше ее вынести в сервис.

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public void unfollowUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
```
Это тоже бизнес логика, вынеси ее тоже в сервис.

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public List<UserDto> getFollowers(long followeeId, UserFilterDto filter) {
+        return subscriptionService.getFollowers(followeeId, filter);
```
Так как ты хочешь вернуть просто сущность, а @Controller этого не умеет, то здесь есть 2 варианта:
- Сделать через ResponseEntity. Которая и будет десериализовать в JSON
- Вместо @Controller, поставить @RestController, который включает в себ яи аннотацию @Controller и аннотацию @ResponseBody - которая как раз и десериализует сущность в JSON

Я за второй вариант )

zaqqaz226 commented on src/main/java/school/faang/user_service/exceptions/DataValidationException.java:
```diff
@@ -0,0 +1,14 @@
+package school.faang.user_service.exceptions;
+
+
+public class DataValidationException extends RuntimeException {
```
Отлично, что решил сделать свое исключение! Так и надо делать.

Commit:
@controller == @restcontroller
валидация followindId == followeeId вынесена в service

zaqqaz226 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,85 @@
+    private final List<UserFilter> userFilters;
+
+    public void followUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
```
Проверка повторяется, так что лучше вынести ее в отдельный приватный метод.

zaqqaz226 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public void unfollowUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
```
Соответственно и тут

zaqqaz226 commented:
Вынеси проверку в отдельный метод и можно писать Unit Tests

Commit:
добавлена проверка isSelfAction
добавлена проверка checkSubscribe

Commit:
Controller Test+
ExperienceFilter Test+
NameFilter Test+
PhoneFilter Test+

Commit:
SubscriptionServiceTest +

Commit:
сокращена логика тестирования testGetFollowers && testGetFollowings
UserNameFilter добавелна проверка на filter.isBlank && username.isBlank - так же покрыто тестами
UserPhoneFilter добавелна проверка на filter.isBlank && phone.isBlank - так же покрыто тестами
